src/raftstore/store/fsm/apply.rs

```diff
@@ -2564,30 +2575,54 @@ impl ApplyFsm {
             .force_send(Msg::LogsUpToDate(region_id));
     }
 
+    #[allow(unused_mut)]
     fn handle_snapshot(&mut self, apply_ctx: &mut ApplyContext, snap_task: GenSnapTask) {
         if self.delegate.pending_remove || self.delegate.stopped {
             return;
         }
+        let mut need_sync =
+            self.delegate.last_sync_apply_index != self.delegate.apply_state.get_applied_index();
+        (|| {
+            fail_point!("apply_on_handle_snapshot_sync", |_| { need_sync = true });
+        })();
+        if need_sync {
+            if apply_ctx.timer.is_none() {
+                apply_ctx.timer = Some(SlowTimer::new());
+            }
+            if apply_ctx.kv_wb.is_none() {
+                apply_ctx.kv_wb = Some(WriteBatch::with_capacity(DEFAULT_APPLY_WB_SIZE));
+            }
+            self.delegate
+                .write_apply_state(&apply_ctx.engines, apply_ctx.kv_wb());
+            fail_point!(
+                "apply_on_handle_snapshot_1_1",
+                self.delegate.id == 1 && self.delegate.region_id() == 1,
+                |_| unimplemented!()
+            );
 
-        // Persists any pending changes of this region.
-        let region_id = self.delegate.region_id();
-        if apply_ctx
-            .apply_res
-            .iter()
-            .any(|res| res.region_id == region_id)
-        {
+            apply_ctx.sync_log_hint = true;
```

Comment:
What if `enable_sync_log` is false?

Reply:
It's okay, apply states are still in the raft cf and they are in the same write-batch as apply results.